firedrake/formmanipulation.py

```diff
@@ -127,6 +128,56 @@ def argument(self, o):
                          for j in numpy.ndindex(V_is[i].value_shape)]
         return self._arg_cache.setdefault(o, as_vector(args))
 
+    def cofunction(self, o):
+        from firedrake import Cofunction, FunctionSpace, MixedFunctionSpace
+        from pyop2 import MixedDat
+        from ufl.classes import ZeroBaseForm
+
+        V = o.function_space()
+
+        # Not on a mixed space, just return ourselves.
+        if len(V) == 1:
+            return o
+
+        # We only need the test space for Cofunction
+        indices = self.blocks[0]
+        V_is = V.subfunctions
+
+        try:
+            indices = tuple(indices)
+            nidx = len(indices)
+        except TypeError:  # Only one index
+            indices = (indices, )
+            nidx = 1
```

Comment:
Can this be
```py
if isinstance(indices, numbers.Integral):
    indices = (indices,)
```
Flow control with exceptions is generally less clear than conditionals.

`nidx` isn't really needed nor I expect is casting `indices = tuple(indices)`.

Reply:
I based the `cofunction` method from the `argument` method so this is left over from there. I'll see what I can take out.

Reply:
Changed to testing `isinstance(indices, int)` and updated the `arguments` method too. `nidx` removed for `len(indices)` because it's only used once.